### DISCUSSION TITLE
archlinux关于neovim python module采用python-pynvim

### DISCUSSION THREAD

**Discussion by popjdh** [ayamir/nvimdots#900] [General]
[archwiki](https://wiki.archlinux.org/title/Neovim)的neovim介绍中有提及[python-pynvim](https://archlinux.org/packages/extra/any/python-pynvim/), 在Prerequisites的介绍中是否可将原有的`pip install neovim --user`修改为`pacman -S python-pynvim`

**Comment by Jint-lzxy** [comment]
Yep but u should use this with caution. The following must be satisfied, otherwise Neovim may fail to find this provider (For most of the time this will not be an issue, but when you somehow mess up your environment, this will be a tough one to debug).
- This is the only `python` installed on your system. Parallel installation is usually not supported _(unless you can guarantee this `python` will always be close to the front of ur `$PATH`, including non-interactive shells)_
- ur `python3`, `pip3` is a link to `python`, `pip`
- All other executables (e.g., `pip`, `pip3`) belong to this `python` package

Why not `pip install neovim --user`? Do you have any concerns?

**Reply by popjdh** [reply]
As I dont originally have `pip` in default env, I use `pacman -S python-pip` to get a `pip`, but when I try `pip install neovim --user`, it outputs:
```plaintext
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try 'pacman -S
    python-xyz', where xyz is the package you are trying to
    install.

    If you wish to install a non-Arch-packaged Python package,
    create a virtual environment using 'python -m venv path/to/venv'.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.

    If you wish to install a non-Arch packaged Python application,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. Make sure you have python-pipx
    installed via pacman.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

**Reply by Jint-lzxy** [reply]
Hmm in this case this should be feasible, as long as you ensure that your package manager only manages one major `python` distribution. You can also install `python-pipx` and use `pipx install neovim` or just bypass this check:
```console
python[3] -m pip install neovim --user --break-system-packages
```

**Reply by popjdh** [reply]
I should using 'pip install neovim -- user' in virtual environments such as pipx or conda (mini conda). 
But for the Python maintained by archLinux, it may be better to use its corresponding maintained package (if it exists).

According to archwiki:
```plaintext
There are several ways to install Python packages on Arch Linux:

Official repositories and AUR — A large number of popular packages are available in the Arch repositories. This is the preferred way to install system-wide packages.

...
```

Can you add a comment prompt for this?

**Reply by Jint-lzxy** [reply]
Thanks for the info! But I seldom use Python since my development target doesn't involve any sort of Python-related stuffs - so for the most precise wording, it's better for someone who understands this language to do this. Sorry I can't do more in this instance.

cc @CharlesChiuGit @ayamir

**Reply by charliie-dev** [reply]
> I should using 'pip install neovim -- user' in virtual environments such as pipx or conda (mini conda).
> But for the Python maintained by archLinux, it may be better to use its corresponding maintained package (if it exists).

sounds like a legit argument to me.
I think it would be nice to add this part to the wiki. Thanks for the heads up!

**Reply by charliie-dev** [reply]
But on the other hand, what's the point of maintaining `python-pip` if u cannot use it to install python pkgs lol

**Reply by popjdh** [reply]
I found a [discussion](https://bbs.archlinux.org/viewtopic.php?id=286788) on system level "Python pip" in Archlinux. 
It seems that manual `pip` may not be suitable for system level Python, even if `pip --user` is used. You need to use `pip --user --break-system-packages`. It doesn't sound safe. 

And for the virtual environments such as pipx or conda (mini conda), you can directly use `pip` without using `--user`.

PS. If you search for 'This environment is externally managed', you can find many cases of Python being marked as external management at the system level.

**Reply by charliie-dev** [reply]
Interesting post, I solve the issue myself in another method, which is installed as many things as possible at user level (or home dir).

For example:
https://github.com/CharlesChiuGit/dotfiles/tree/main/install_scripts

**Reply by popjdh** [reply]
@CharlesChiuGit  It seems that you are installing and managing most of the dependencies through your own scripts (under the "app" and "update" folders). 

For systems such as ubuntu, it may be good, but for systems like archLinux with rolling updates, it may be a bit troublesome. You need to add an additional pacman hook to synchronize updates in the system update.

**Reply by charliie-dev** [reply]
no need lol
the script i wrote are mostly disentangled from the system pkgs. The initial idea of creating those scripts is b/c ubuntu updates it's repo too slow, sometimes i need the bleeding-edge releases directly from the github, installing those apps in user space is just a good side effect. 

not to mention most apps i use is written in golang or rust, their toolchain is pretty self-sustaining; therefore disentangle themselves from system pkg even further and will work on most linux distros, macos and windows.

only some c-based apps, like tmux, need some libs from the system pkg, but their very few and i can manually deal with that.

**Reply by popjdh** [reply]
It sounds like a good design. On archLinux, you can try to build a PKGBUILD for direct maintenance using pacman.

**Reply by charliie-dev** [reply]
u could do that, but it won't be cross-distro or cross-os.

**Comment by ayamir** [comment]
You are right. For archlinux, it's better to install python module by `pacman`. I have edited the wiki.

**Reply by popjdh** [reply]
I need to correct it, something like `pip install neovim` is still necessary.
For different Python project, users may use system level Python or Python in a virtual environment, so the "neovim" package needs to be installed in each environment.

- For system level Python, you need to use `pacman -S python-pynvim`.
- For virtual environment, you should use corresponding installation command. For example, installing this package under Conda should use `conda install -c conda-forge neovim`

Therefore, it may be necessary to point out that **users need to choose the installation method based on the environment they are using**, and provide some examples of mainstream environment(system python, pipx, conda, etc).

**Comment by charliie-dev** [comment]
https://youtu.be/35PQrzG0rG4

Seems like it's not Arch's own problem.
